linear_operator/operators/permutation_linear_operator.py

```diff
@@ -0,0 +1,168 @@
+    Upon left-multiplication, it permutes the first non-batch dimension of a tensor.
+
+    Args:
+        - perm: A permutation tensor with which to permute the first non-batch
+            dimension through matmul. Should have integer elements.
+        - inv_perm: Optional tensor representing the inverse of perm, is computed
+            via a O(n log(n)) sort if not given.
+        - validate_args: Boolean
+    """
+
+    def __init__(
+        self,
+        perm: Tensor,
+        inv_perm: Optional[Tensor] = None,
+        validate_args: bool = True,
+    ):
+        if not isinstance(perm, Tensor):
+            raise ValueError("perm is not a Tensor.")
+
+        if inv_perm is not None:
+            if perm.shape != inv_perm.shape:
+                raise ValueError("inv_perm does not have the same shape as perm.")
+
+            batch_indices = self._batch_indexing_helper(perm.shape[:-1])
+            sorted_perm = perm[batch_indices + (inv_perm,)]
+        else:
+            sorted_perm, inv_perm = perm.sort(dim=-1)
+
+        if validate_args:
+            if torch.is_floating_point(sorted_perm) or torch.is_complex(sorted_perm):
+                raise ValueError("perm does not have integer elements.")
+
+            for i, si in enumerate(sorted_perm):
+                if (sorted_perm[..., i] != i).any():
+                    raise ValueError(
+                        f"Invalid perm-inv_perm input, index {i} missing or not at "
+                        f"correct index for permutation with {perm.shape = }."
+                    )
+
+        self.perm = perm
+        self.inv_perm = inv_perm
+        super().__init__(perm, inv_perm, validate_args=validate_args)
+
+    def _matmul(self, rhs: Tensor) -> Tensor:
+        # input rhs is guaranteed to be at least two-dimensional due to matmul implementation
+        self._matmul_check_shape(rhs)
+
+        # batch broadcasting logic
+        batch_shape = self._matmul_batch_shape(rhs)
+        expanded_rhs = rhs.expand(*batch_shape, *rhs.shape[-2:])
+        ndim = expanded_rhs.ndim
+
+        batch_indices = self._batch_indexing_helper(batch_shape)
+        batch_indices = tuple(index[..., None] for index in batch_indices)  # expanding to non-batch dimensions
+        perm_indices = self.perm[..., None]
+        final_indices = torch.arange(rhs.shape[-1])[(None,) * (ndim - 1) + (slice(None),)]
+        indices = batch_indices + (perm_indices, final_indices)
+        return expanded_rhs[indices]
+
+    def _batch_indexing_helper(self, batch_shape: torch.Size) -> Tuple:
+        indices = []
+        for i, n in enumerate(batch_shape):
+            indices.append(
+                torch.arange(n)[
+                    (None,) * i
+                    + (slice(None),)
+                    + (None,) * (len(batch_shape) - i - 1)
+                    + (None,)  # adding one non-batch dimension
+                ]
+            )
+        return tuple(indices)
```

Comment:
Can make a tuple right away. Also could use `view` here instead (not necessary though)
```suggestion
        return tuple(
            torch.arange(n)[
                (None,) * i
                + (slice(None),)
                + (None,) * (len(batch_shape) - i - 1)
                + (None,)  # adding one non-batch dimension
            ]
            for i, n in enumerate(batch_shape)
        )
```